BLE Indoor Positioning/src/main/java/com/nexenio/bleindoorpositioning/location/distance/BeaconDistanceCalculator.java

```diff
@@ -23,6 +23,26 @@ public static float calculateDistanceTo(Beacon beacon) {
         return calculateDistanceTo(beacon, beacon.getFilteredRssi());
     }
 
+    /**
+     * Use Pythagoras with altitude to calculate more accurate distance
+     */
+    public static float calculateDistanceWithHeightTo(Beacon beacon, float rssi) {
+        double altitude = beacon.getLocation().getAltitude();
+        // check for altitude
+        if (altitude > 0) {
+            double distance = calculateDistanceTo(beacon, rssi);
+            double delta = Math.pow(distance, 2) - Math.pow(altitude, 2);
+            // check for negative square root
+            if (delta < 0) {
```

Comment:
If the measured distance is less than the altitude delta, the device must be located directly below the beacon. No point to use pythagoras in this case.